service.py

```diff
@@ -0,0 +1,263 @@
+
+    This method is a manager that handles methods around a HathiRecord object.
+    Each method creates/enhances a part of the SFR metadata object, allowing
+    for the object to both be built up and its components easily treated
+    as seperate components if necessary
+
+    Arguments:
+    row -- list of fields from the HathiTrust source CSV file
+    columns -- list of columns that corresponds to the source row
+    countryCodes -- dict of country code and name translations
+
+    Output: None, writes resulting work record to a Kinesis stream
+    """
+
+    logger.info('Reading entry for HathiTrust item {}'.format(row[0]))
+
+    logger.debug('Generating source dict from row and column names')
+    # This quickly builds a dictionary with column names that can be used to
+    # retrieve specific values
+    hathiDict = dict(zip(columns, row))
+    # Generate a hathi record object with the source dict
+    hathiRec = HathiRecord(hathiDict)
+
+    try:
+        logger.debug('Generating work record for bib record {}'.format(
+            hathiRec.ingest['bib_key']
+        ))
+        hathiRec.buildWork()
+
+        logger.debug('Generating instance record for hathi record {}'.format(
+            hathiRec.ingest['htid']
+        ))
+        hathiRec.buildInstance(countryCodes)
+        hathiRec.work.instances.append(hathiRec.instance)
+
+        logger.debug('Generating an item record for hathi record {}'.format(
+            hathiRec.ingest['htid']
+        ))
+        hathiRec.buildItem()
+        hathiRec.instance.formats.append(hathiRec.item)
+
+        logger.debug('Generate a rights object for the associated rights statement {}'.format(
+            hathiRec.ingest['rights']
+        ))
+        # Generate a stand-alone rights object that contains the hathi
+        # generated rights information
+        hathiRec.createRights()
+        # At present these rights are assigned to all three levels in the SFR
+        # model work, instance and item. While this data certainly pertains to
+        # the instance and item records retrieved here, its relevance is
+        # unclear for the work record. It will be possible to have conflicting
+        # rights statements for works and instances
+        hathiRec.work.rights = [hathiRec.rights]
+        hathiRec.instance.rights = [hathiRec.rights]
+        hathiRec.item.rights = [hathiRec.rights]
+    except DataError as err:
+        logger.error('Unable to process record {}'.format(
+            hathiRec.ingest['htid']
+        ))
+        logger.debug(err.message)
+        raise ProcessingError('DataError', err.message)
+
+    try:
+        logger.debug('Writing hathi record {} to kinesis for ingest'.format(
+            hathiRec.work.primary_identifier.identifier
+        ))
+        KinesisOutput.putRecord({
+            'status': 200,
+            'type': 'work',
+            'method': 'insert',
+            'data': hathiRec.work
```

Comment:
Sorry if this is a dumb question. The work is put into Kinesis -- does anything happen with the instance or item?

Reply:
The instance and item are added as children record to the work as part of the `parseRow` method, so they are included here.